text/0000-portability-lint.md

```diff
@@ -0,0 +1,574 @@
+
+**What does it mean for a portability to be narrower?** In general, portability
+is a logical expression, using the operators `all`, `any`, `not` on top of
+primitive expressions like `unix`. Portability `P` is narrower than portability
+`Q` if `P` *implies* `Q` as a logic formula.
+
+In general, comparing two portabilities is equivalent to solving SAT, an
+NP-complete problem -- a frightening prospect for a lint! However, in practice
+these comparisons are going to be relatively small and simple SAT instances,
+given the way that `cfg` is used. We can likely get away with a naive SAT
+implementation, perhaps with a handful of optimiziations specific to our
+use-case. In the limit, there are also many well-known techniques for solving
+SAT efficiently even on very large examples that arise in real-world usage.
+
+#### Known implications
+
+Another aspect of portability comparison is the relationship between things like
+`unix` and `linux`. In logical terms, we want to say that `linux` implies
+`unix`, for example.
+
+The primitive portabilities we'll be comparing are all *built in* (since we are
+not including Cargo features). We can thus build in implications between these
+built-in portabilities, which are fed to the SAT solver when comparing
+portabilities. The end result is that code like the following should pass the lint:
+
+```rust
+#[cfg(unix)]
+fn unix_only() { .. }
+
+#[cfg(linux)]
+fn linux_only() {
+    // permitted since `linux` implies `unix`
+    unix_only()
+}
+```
+
+### Determining the portability of referenced items
+
+**How is the portability of a referenced item determined?** The lint will simply
+resolve an item to its definition, and use the portability of that
+definition. In particular, when invoking trait items, the portability will be
+that of the item's *definition*, rather than that of the `impl`. See
+[Unresolved questions][unresolved] for more on this point; it's something that
+could be revised over time.
+
+#### `match_cfg`
+
+The `match_cfg` macro takes a sequence of `cfg` patterns, followed by `=>` and
+an expression. Its syntax and semantics resembles that of `match`. However,
+there are some special considerations when checking portability:
+
+* When descending into an arm of a `match_cfg`, the arm is checked against
+  portability that includes the pattern for the arm.
+
+* The portability for the `match_cfg` itself is understood as `any(p1, ...,
+  p_n)` where the `match_cfg` patterns are `p1` through `p_n`.
+
+Thus, for example, the following code will pass the lint:
+
+```rust
+#[cfg(windows)]
+fn windows_only() { .. }
+
+#[cfg(unix)]
+fn unix_only() { .. }
+
+#[cfg(any(windows, unix))]
+fn portable() {
+    // the expression here has portability `any(windows, unix)`
+    match_cfg! {
+        windows => {
+            // allowed because we are within a scope with
+            // portability `all(any(windows, unix), windows)`
+            windows_only()
+        }
+        unix => {
```

Comment:
It would be reassuring to see a `match_cfg` example with more complex CFGs. So far no examples of `match_cfg` in conjunction with value-cfg (e.g. `target_os = "linux"`) have been provided.